Large pull request to bring back the gazebo plugins

jwhite66 opened the pull request:
This pull request includes a series of patches that brings the gazebo plugins back into the wpilib tree and builds them on linux, when the appropriate tools are available.

This includes older gradle scripts that would likely be best revised by someone expert in the new configuration.  However, the bones should be solid, and this should restore the gazebo plugin functionality (which is one part of the puzzle).

frcjenkins commented:
Can one of the admins verify this patch?

frcjenkins commented:
Can one of the admins verify this patch?

frcjenkins commented:
Can one of the admins verify this patch?

frcjenkins commented:
Can one of the admins verify this patch?

jwhite66 commented:
It worked fine 19 days ago, and the only conflict that has arisen looks minor.  I have further branches I'm waiting to transmit that make it more elegant to use, and my plan is after those are in to turn to the Gradle RIO tree to make it possible for teams to use it.

calcmogul commented on hal/src/main/native/include/MockData/EncoderData.h:
```diff
@@ -17,6 +17,7 @@ extern "C" {
 #endif
 
 void HALSIM_ResetEncoderData(int32_t index);
+int16_t HALSIM_GetDigitalChannelA(int index);
```
HAL things use int32_t instead of int.

jwhite66 commented on hal/src/main/native/sim/MockData/EncoderDataInternal.h:
```diff
@@ -19,6 +19,9 @@
 class EncoderData {
  public:
+  void SetDigitalChannelA(int16_t channel);
+  int16_t GetDigitalChannelA();
```
If I understand the question correctly, I would hesitate because the remainder of that class (and, in fact, most of the hal code) does not use the trailing const.  I would think it better, if that semantic is desired, to introduce it as a patch that changes the class as a whole.

calcmogul replied:
Member functions that do not mutate the class's internal state (like getters) should be const-qualified. Since nothing overrides that function and it was added in this patch series, there's no reason not to const-qualify it now.

jwhite66 replied:
Forgive me, I think I've taken offense where none was meant.  And on a trivial matter, what's worse.  But I feel that  "no reason" dismisses my point and belittles it.  Let me restate it to make sure I was clear.

I believe you are proposing that I introduce a relatively obscure semantic into a body of code in an exceptional way, which I believe would obfuscate the code.  (i.e. why on earth is GetDigitalChannelA() the *only* one that is const qualified?  What's so different about it?)

Now I will make whatever change you request.  But I wanted to be sure you heard my concern and that you believe the benefit outweighs my concern.  You can call my concern a 'tiny reason' if it helps :-).

I would also defer to a request on your part that I submit a patch const qualifying all the gettors of the class.

calcmogul replied:
I'm sorry if I sound patronizing, but constness is not an obscure semantic; it's a fundamental feature commonly used in idiomatic C++. It's a way to ensure at compile time that a function does not have side effects, which the user of the function can notice by looking at the function signature and rely on. Here's the relevant C++ core guideline: https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rconst-fct. Member functions should be const-qualified unless they mutate internal state. We use this all over the place in the rest of WPILib like `double SpeedController::Get() const`. It's enforcing through the compiler what the API writer already intended (the `override` keyword is another example of this).

Of course, I'm assuming the API user is familiar with the feature, but that should be a good assumption since it's a basic feature covered in basically every C++ book I've seen (I use _Programming: Principles and Practice Using C++, 2nd Ed._ to teach my students, and 9.7.4 of the chapter on classes covers this feature). Appending `const` to the function signature is strictly a benefit because it gives the user more information (namely that the function doesn't cause side effects). This doesn't limit where the user can use the function. In fact, it also allows them to use the function in more places like within other const-qualified functions.

Const-qualifying a function is like a stronger variant of marking a Java variable `final`. You don't have to do it, but you should when you can. EncoderDataInternal.h's other functions don't follow this guideline and need to be fixed (but that can happen in a separate PR).

jwhite66 commented:
Yeah, the frcgazebo models add 156M of data that is going to be very rarely used; I see good rationale for parking them elsewhere.  It's mildly tricky, as it's just the meshes that are large, and the model sdf files work nicely in git, so it'd be good to have them in a git tree somewhere.  Not sure what this implies for this PR; if I should pull it and wait, or if I should attempt to populate a git tree elsewhere that would be suitable for the models, and then revise the PR to fit.

ThadHouse commented:
I think for now, can we get the gazebo model files removed, and maybe you can zip them and email them to me? I think we do want to do them externally. Since this doesn't currently build, we can get the the building and everything implemented later.

jwhite66 commented:
I've zipped the frcgazebo directory and stashed it here:  https://drive.google.com/open?id=13KEe_kOftxCFJZJOl1UhENPnyWfatVVj.  The large files are the .STL files that provide the visual meshes.

PeterJohnson commented:
Can you amend this PR to remove the STL files from the PR and fix the conflict in .styleguide?  I think we can merge it then.

jwhite66 commented:
I removed all of the meshes (mostly STL files, but a few large .dae files as well), and adjusted the styleguide until I was able to merge into master.  Note that this leaves this patch not usable as described in the README; I'd like a chance to add a patch to fix that once the official hosting location for the meshes is available.

PeterJohnson commented:
Should all of the models/worlds move too?  Not sure how big those are.  Also, travis failed with formatting stuff, should be fairly easy to fix.

jwhite66 commented:
The models and worlds should be somewhere under git control, as they are, essentially, code.  In an ideal world, they will also be getting some fairly heavy revision to tune the models and world to be useful for sample programming.  So my preference is that they be in a git tree somewhere.  It doesn't have to be this tree, and yes, they really do logically belong with the meshes.
A puzzling note on the Travis error:  I can't get a locally invoked copy of wpiformat to generate that error.  I presume it's operator error, but I've been fairly studious to follow the travis script, and I get a different result.

calcmogul commented on .styleguide:
```diff
@@ -10,6 +10,8 @@ cppSrcFileInclude {
 
 generatedFileExclude {
   gtest/
+  simulation/gz_msgs/src/include/simulation/gz_msgs/msgs.h
```
Please lexographically sort these.

calcmogul replied:
Also, .styleguide uses regular expressions. This should be `simulation/gz_msgs/src/include/simulation/gz_msgs/msgs\.h$`.

jwhite66 commented:
Okay, dug in a bit more and figured it out.   I had a subdirectory with a test program binary which was causing wpiformat to not only generate a UTF-8 complaint, but to crash, and not proceed to further directories. So I thought I was running wpiformat, but I was not running it completely.
As far as file names, I generally preserved existing file names when I restored them.  Thus, the frc_gazebo_plugin files are largely snake case.  When I added new files, I attempted to follow what I understood to be correct convention, hence the halsim_gazebo CamelCase files.  Arguably, I got the 'clockTest' directory and file wrong; it is neither snake or camel.  I am happy to rename it to either, if one is clearly more correct than the other.
On the .styleguide, Peter had requested that I fix it up so that it would merge.  I could not find a single commit that would achieve that within this PR and preserve lexical order. I think I would need to submit a new PR in order to preserve lexical order and allow it to merge without conflict.  I'm happy to do that, or to take a clue bat on a surgical commit I can make to resolve the merge conflict without breaking lexical order.

calcmogul commented:
I mean, the right way to fix that conflict is to do a rebase against master. I just ran it locally and the fixes for conflicts along the way are trivial.

calcmogul commented on .styleguide:
```diff
@@ -13,6 +13,8 @@ generatedFileExclude {
   ni-libraries/lib/
   FRCNetComm\.java$
+  simulation/frc_gazebo_plugins/frcgazebo
+  simulation/gz_msgs/src/include/simulation/gz_msgs/msgs\.h
```
This should have $ appended so it matches msgs.h and nothing else.

calcmogul commented on .styleguide:
```diff
@@ -13,6 +13,8 @@ generatedFileExclude {
   ni-libraries/include/
   ni-libraries/lib/
   FRCNetComm\.java$
+  simulation/frc_gazebo_plugins/frcgazebo
```
This should have / appended so it matches just a directory named "frcgazebo" and nothing else.

PeterJohnson commented:
@frcjenkins retest this please

PeterJohnson commented:
@frcjenkins retest this please

PeterJohnson commented:
@jwhite66 can you remove the models and jpg/png files?  As you stated they logically belong with the meshes.  I feel they don't belong in this PR as they're really assets rather than code.

PeterJohnson commented:
Thank you for your hard work on this!
